src/Dapr.Messaging/PublishSubscribe/PublishSubscribeReceiver.cs

```diff
@@ -47,159 +42,124 @@ internal sealed class PublishSubscribeReceiver : IAsyncDisposable
+    /// <param name="handler">The delegate reflecting the action to take upon messages received by the subscription.</param>
+    internal PublishSubscribeReceiver(string pubSubName, string topicName, DaprSubscriptionOptions options, ConnectionManager connectionManager, TopicMessageHandler handler)
     {
-        this.pubsubName = pubsubName;
+        this.pubSubName = pubSubName;
         this.topicName = topicName;
         this.options = options;
-        connectionManager = new ConnectionManager(daprClient);
+        this.connectionManager = connectionManager;
+        this.messageHandler = handler;
     }
 
     /// <summary>
     /// Dynamically subscribes to messages on a PubSub topic provided by the Dapr sidecar.
     /// </summary>
     /// <param name="cancellationToken">Cancellation token.</param>
     /// <returns>An <see cref="IAsyncEnumerable{TopicMessage}"/> containing messages provided by the sidecar.</returns>
-    public IAsyncEnumerable<TopicMessage> SubscribeAsync(CancellationToken cancellationToken)
-    {
-        _ = FetchDataFromSidecar(channel.Writer, cancellationToken);
-        return ReadMessagesFromChannelAsync(channel.Reader, cancellationToken);
-    }
-
-    /// <summary>
-    /// Specifies the action that should be taken on the message after processing it.
-    /// </summary>
-    /// <param name="messageId">The identifier of the message to acknowledge.</param>
-    /// <param name="messageAction">The action to take on the message.</param>
-    /// <param name="cancellationToken">Cancellation token.</param>
-    public async Task AcknowledgeMessageAsync(string messageId, TopicMessageAction messageAction, CancellationToken cancellationToken)
+    public async Task SubscribeAsync(CancellationToken cancellationToken)
     {
         var stream = await connectionManager.GetStreamAsync(cancellationToken);
-        await stream.RequestStream.WriteAsync(new P.SubscribeTopicEventsRequestAlpha1
+        //Retrieve the messages from the sidecar and write to the channel
+        _ = FetchDataFromSidecarAsync(stream, channel.Writer, cancellationToken);
+        
+        //Read the messages one-by-one out of the channel
+        try
         {
-            EventResponse = new P.SubscribeTopicEventsResponseAlpha1
+            while (await channel.Reader.WaitToReadAsync(cancellationToken))
             {
-                Id = messageId,
-                Status = new C.TopicEventResponse
+                while (channel.Reader.TryRead(out var message))
                 {
-                    Status = messageAction switch
+                    using var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+                    cts.CancelAfter(options.MessageHandlingPolicy.TimeoutDuration);
+
+                    //Evaluate the message and return an acknowledgement result
+                    var messageAction = await messageHandler(message, cts.Token);
```

Comment:
So with this change, acknowledgments are sent back to the sidecar in the same order in which their corresponding message was received, correct?  I thinks that's fine--but I don't think we *have* to do that, necessarily.  (Just to be clear, while we need to write to the stream sequentially, I don't think the sidecar requires they be written in any particular order.  One option would be to have a separate `Channel` for acks that is written to as soon as the message handling is complete (or times out).  Then a separate `Task` can sit and wait on the channel, writing to the gRPC stream as they come in.  That would allow messages to be ack'd as soon as possible while also ensuring acks are written to the stream one-by-one.

There may be a larger question of whether it makes sense for an application to potentially have multiple messages being handled at the same time, or whether it would prefer messages to be handled one-by-one.

Reply:
I confirmed with @yaron2 that the message acknowledgement order doesn't matter. I was looking through the Go implementation and it doesn't seem to have any restriction on there being multiple subscribers to the same combination - it just kicks off yet another subscription. I don't see any reason we can't do that ourselves then and, as you suggested, just have them all write to a Channel and have a single task read off it to handles the acks. I'll have an update with this soon.